### ISSUE TITLE
Example of TFMarianMTModel is not working very well

### ISSUE THREAD

**Issue by Zhenzi-Weng** [huggingface/transformers#17357]
Hello, I copied the TFMarianMTModel example code from [https://huggingface.co/docs/transformers/v4.19.2/en/model_doc/marian#transformers.TFMarianMTModel](url) and compiled in Tensorflow 2.4.1. But it turned out the output of "model.generate(**batch)" is always 512.

`from transformers import MarianTokenizer, TFMarianMTModel
from typing import List

src = "fr"  # source language
trg = "en"  # target language
sample_text = "où est l'arrêt de bus ?"
model_name = f"Helsinki-NLP/opus-mt-{src}-{trg}"

model = TFMarianMTModel.from_pretrained(model_name)
tokenizer = MarianTokenizer.from_pretrained(model_name)
batch = tokenizer([sample_text], return_tensors="tf")
gen = model.generate(**batch)
tokenizer.batch_decode(gen, skip_special_tokens=True)`

Could you help with it?

@patrickvonplaten

**Comment by patrickvonplaten** [comment]
Hey @Zhenzi-Weng, 

When I execute the code above, I'm getting an error. I'm assuming you want to execute the following:

```python
from transformers import MarianTokenizer, TFMarianMTModel
from typing import List

src = "fr"  # source language
trg = "en"  # target language
sample_text = "où est l'arrêt de bus ?"
model_name = f"Helsinki-NLP/opus-mt-{src}-{trg}"

model = TFMarianMTModel.from_pretrained(model_name, from_pt=True)
tokenizer = MarianTokenizer.from_pretrained(model_name)
batch = tokenizer([sample_text], return_tensors="tf")
gen = model.generate(**batch)
tokenizer.batch_decode(gen, skip_special_tokens=True)
```

**Comment by patrickvonplaten** [comment]
When executing the above code, I'm getting a pretty reasonable answer: 

```
["Where's the bus stop?"]   
```

=> What is the problem here exactly?

**Comment by Zhenzi-Weng** [comment]
> When executing the above code, I'm getting a pretty reasonable answer:
> 
> ```
> ["Where's the bus stop?"]   
> ```
> 
> => What is the problem here exactly?

Hello @patrickvonplaten , thanks for your reply. I tried your code and it did output a reasonable translation, but I believe your code is for Pytorch. I‘m using Tensorflow and the problem is the output will contain many invalid characters. Would you mind to try the following code again? Many thanks. My Tensorflow version is 2.4.1 and I installed transformers v4.19.0.
`from transformers import MarianTokenizer, TFMarianMTModel

src = "en"  # source language
trg = "zh"  # target language

model_name = f"Helsinki-NLP/opus-mt-{src}-{trg}"
model = TFMarianMTModel.from_pretrained(model_name)
tokenizer = MarianTokenizer.from_pretrained(model_name)

sample_text = "My name is Wolfgang and I live in Berlin"
batch = tokenizer([sample_text], return_tensors="tf")

generated_ids = model.generate(**batch)
tokenizer.batch_decode(generated_ids, skip_special_tokens=True)`

**Comment by patrickvonplaten** [comment]
Hey @Zhenzi-Weng,

Note that my codesnippet above is for Tensorflow, the model is loaded into a `TFMarianMTModel` class. We just need to have PyTorch installed to convert the PT weights to TF before

**Comment by jamie0725** [comment]
Hi @patrickvonplaten, related to this question, why is it that the TF version of MarianMTModel much slower than the PT version? Maybe it has something to do with the length of generating output?

**Comment by patrickvonplaten** [comment]
TF is generally much slower in eager mode than PyTorch - you could try to compile the generate method in Tensorflow which should speed things up :-) 

cc @JoaoLages

**Comment by gante** [comment]
@jamie0725 correct, TF (eager execution) is much slower than PyTorch. However, with XLA, it should be much faster. Have a look at [this blog post](https://huggingface.co/blog/tf-xla-generate)!
